compiler/src/vm/heap/object.rs

```diff
@@ -76,36 +76,32 @@ impl Struct {
     /// Otherwise, returns the index of where the key would be inserted.
     fn index_of_key(&self, heap: &Heap, key: Pointer) -> Result<usize, usize> {
         let hash = key.hash(heap);
-        let index_of_hash = self
+        let index_of_first_hash_occurrence = self
             .fields
-            .binary_search_by_key(&hash, |(hash, _, _)| *hash)?;
-        let going_right = self.fields[index_of_hash..]
+            .partition_point(|(existing_hash, _, _)| *existing_hash < hash);
+        let fields_with_same_hash = self.fields[index_of_first_hash_occurrence..]
             .iter()
             .enumerate()
             .take_while(|(_, (existing_hash, _, _))| *existing_hash == hash)
-            .map(|(i, (_, key, _))| (index_of_hash + i, key));
-        let going_left = self.fields[0..index_of_hash]
-            .iter()
-            .rev()
-            .enumerate()
-            .take_while(|(_, (existing_hash, _, _))| *existing_hash == hash)
-            .map(|(i, (_, key, _))| (index_of_hash - i - 1, key));
+            .map(|(i, (_, key, _))| (index_of_first_hash_occurrence + i, key));
 
-        for (index, existing_key) in going_left.chain(going_right) {
+        for (index, existing_key) in fields_with_same_hash {
```

Comment:
We could even perform a binary search within entries with the same hash in case the key has the same pointer, but I don't think that would be worth it. Also, we'd have to sort the entries by key pointers as well